
<!DOCTYPE html>

<html lang="en">
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="generator" content="Docutils 0.18.1: http://docutils.sourceforge.net/" />

    <title>Specifying genomes &#8212; pybedtools 0.9.0 documentation</title>
    <link rel="stylesheet" type="text/css" href="_static/pygments.css" />
    <link rel="stylesheet" type="text/css" href="_static/alabaster.css" />
    <script data-url_root="./" id="documentation_options" src="_static/documentation_options.js"></script>
    <script src="_static/jquery.js"></script>
    <script src="_static/underscore.js"></script>
    <script src="_static/_sphinx_javascript_frameworks_compat.js"></script>
    <script src="_static/doctools.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Randomization" href="topical-random.html" />
    <link rel="prev" title="Notes on BAM file semantics" href="topical-bam-semantics.html" />
   
  <link rel="stylesheet" href="_static/custom.css" type="text/css" />
  
  
  <meta name="viewport" content="width=device-width, initial-scale=0.9, maximum-scale=0.9" />

  </head><body>
  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          

          <div class="body" role="main">
            
  <section id="specifying-genomes">
<span id="genomes"></span><h1>Specifying genomes<a class="headerlink" href="#specifying-genomes" title="Permalink to this heading">¶</a></h1>
<p>This section illustrates the use of genome files for use with BEDTools
programs that need to know chromosome limits to prevent out-of-range
coordinates.</p>
<p>Using BEDTools programs like <code class="file docutils literal notranslate"><span class="pre">slopBed</span></code> or <code class="file docutils literal notranslate"><span class="pre">shuffleBed</span></code>  from the command
line requires “genome” or “chromsizes” files.  <code class="xref py py-mod docutils literal notranslate"><span class="pre">pybedtools</span></code> comes with
common genome assemblies already set up as a dictionary with chromosomes as
keys and zero-based (start, stop) tuples as values:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">pybedtools</span> <span class="kn">import</span> <span class="n">genome_registry</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">genome_registry</span><span class="o">.</span><span class="n">dm3</span><span class="p">[</span><span class="s1">&#39;chr2L&#39;</span><span class="p">]</span>
<span class="go">(0, 23011544)</span>
</pre></div>
</div>
<p>The rules for specifying a genome for methods that require a genome are as
follows (use whatever is most convenient):</p>
<ul class="simple">
<li><p>Use <code class="file docutils literal notranslate"><span class="pre">g</span></code> to specify either a filename or a dictionary</p></li>
<li><p>Use <code class="file docutils literal notranslate"><span class="pre">genome</span></code> to specify either an assembly name or a dictionary</p></li>
</ul>
<p>Below are examples of each.</p>
<section id="as-a-file">
<h2>As a file<a class="headerlink" href="#as-a-file" title="Permalink to this heading">¶</a></h2>
<p>This is the typical way of using BEDTools programs, by specifying an existing genome
file with <code class="file docutils literal notranslate"><span class="pre">g</span></code>:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">a</span> <span class="o">=</span> <span class="n">pybedtools</span><span class="o">.</span><span class="n">example_bedtool</span><span class="p">(</span><span class="s1">&#39;a.bed&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">b</span> <span class="o">=</span> <span class="n">a</span><span class="o">.</span><span class="n">slop</span><span class="p">(</span><span class="n">b</span><span class="o">=</span><span class="mi">100</span><span class="p">,</span> <span class="n">g</span><span class="o">=</span><span class="s1">&#39;hg19.genome&#39;</span><span class="p">)</span>
</pre></div>
</div>
</section>
<section id="as-a-string">
<h2>As a string<a class="headerlink" href="#as-a-string" title="Permalink to this heading">¶</a></h2>
<p>This is probably the most convenient way of specifying a genome.  If the
genome exists in the genome registry it will be used directly.  Alternatively,
if you have <a href="#id1"><span class="problematic" id="id2">`genomepy&lt;https://github.com/vanheeringen-lab/genomepy/&gt;`_</span></a>
installed, you can use the genomepy genome name, such as <code class="file docutils literal notranslate"><span class="pre">hg38</span></code>.  In this case,
the genome file will be located automatically.  Finally, if the genome is not
in the registry or managed by genomepy, it will automatically be downloaded
from UCSC.  You must use the <code class="file docutils literal notranslate"><span class="pre">genome</span></code> kwarg for this; if you use <code class="file docutils literal notranslate"><span class="pre">g</span></code> a string
will be interpreted as a filename:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">c</span> <span class="o">=</span> <span class="n">a</span><span class="o">.</span><span class="n">slop</span><span class="p">(</span><span class="n">b</span><span class="o">=</span><span class="mi">100</span><span class="p">,</span> <span class="n">genome</span><span class="o">=</span><span class="s1">&#39;hg19&#39;</span><span class="p">)</span>
</pre></div>
</div>
</section>
<section id="as-a-dictionary">
<h2>As a dictionary<a class="headerlink" href="#as-a-dictionary" title="Permalink to this heading">¶</a></h2>
<p>This is a good way of providing custom coordinates; either <code class="file docutils literal notranslate"><span class="pre">g</span></code> or <code class="file docutils literal notranslate"><span class="pre">genome</span></code>
will accept a dictionary:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">d</span> <span class="o">=</span> <span class="n">a</span><span class="o">.</span><span class="n">slop</span><span class="p">(</span><span class="n">b</span><span class="o">=</span><span class="mi">100</span><span class="p">,</span> <span class="n">g</span><span class="o">=</span><span class="p">{</span><span class="s1">&#39;chr1&#39;</span><span class="p">:(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">10000</span><span class="p">)})</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">e</span> <span class="o">=</span> <span class="n">a</span><span class="o">.</span><span class="n">slop</span><span class="p">(</span><span class="n">b</span><span class="o">=</span><span class="mi">100</span><span class="p">,</span> <span class="n">genome</span><span class="o">=</span><span class="p">{</span><span class="s1">&#39;chr1&#39;</span><span class="p">:(</span><span class="mi">1</span><span class="p">,</span><span class="mi">100000</span><span class="p">)})</span>
</pre></div>
</div>
<p>Make sure that all these different methods return the same results</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">b</span> <span class="o">==</span> <span class="n">c</span> <span class="o">==</span> <span class="n">d</span> <span class="o">==</span> <span class="n">e</span>
<span class="go">True</span>
</pre></div>
</div>
</section>
<section id="converting-to-a-file">
<h2>Converting to a file<a class="headerlink" href="#converting-to-a-file" title="Permalink to this heading">¶</a></h2>
<p>Since BEDTools programs operate on files, the fastest choice will be to
use an existing file.  While the time to convert a dictionary to a file is
extremely small, over 1000’s of files (e.g., for Monte Carlo simulations),
the time may add up.  The function <code class="xref py py-func docutils literal notranslate"><span class="pre">pybedtools.chromsizes_to_file()</span></code>
will create a file from a dictionary or string:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="c1"># with no filename specified, a tempfile will be created</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pybedtools</span><span class="o">.</span><span class="n">chromsizes_to_file</span><span class="p">(</span><span class="n">pybedtools</span><span class="o">.</span><span class="n">chromsizes</span><span class="p">(</span><span class="s1">&#39;dm3&#39;</span><span class="p">),</span> <span class="s1">&#39;dm3.genome&#39;</span><span class="p">)</span>
<span class="go">&#39;dm3.genome&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="nb">open</span><span class="p">(</span><span class="s1">&#39;dm3.genome&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">read</span><span class="p">())</span>
<span class="go">chr2L       23011544</span>
<span class="go">chr2LHet    368872</span>
<span class="go">chr2R       21146708</span>
<span class="go">chr2RHet    3288761</span>
<span class="go">chr3L       24543557</span>
<span class="go">chr3LHet    2555491</span>
<span class="go">chr3R       27905053</span>
<span class="go">chr3RHet    2517507</span>
<span class="go">chr4        1351857</span>
<span class="go">chrM        19517</span>
<span class="go">chrU        10049037</span>
<span class="go">chrUextra   29004656</span>
<span class="go">chrX        22422827</span>
<span class="go">chrXHet     204112</span>
<span class="go">chrYHet     347038</span>
</pre></div>
</div>
</section>
</section>


          </div>
          
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
<h1 class="logo"><a href="index.html">pybedtools</a></h1>








<h3>Navigation</h3>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="main.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="main.html#running-tests-compiling-docs">Running tests, compiling docs</a></li>
<li class="toctree-l1"><a class="reference internal" href="3-brief-examples.html">Three brief examples</a></li>
<li class="toctree-l1"><a class="reference internal" href="tutorial-contents.html">Tutorial Contents</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="topical-documentation-contents.html">Topical Documentation</a><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="topical-design-principles.html">Design principles</a></li>
<li class="toctree-l2"><a class="reference internal" href="topical-create-a-bedtool.html">Creating a <code class="xref py py-class docutils literal notranslate"><span class="pre">BedTool</span></code></a></li>
<li class="toctree-l2"><a class="reference internal" href="topical-saving.html">Saving <code class="xref py py-class docutils literal notranslate"><span class="pre">BedTool</span></code> results</a></li>
<li class="toctree-l2"><a class="reference internal" href="topical-iterators.html">Using BedTool objects as iterators/generators</a></li>
<li class="toctree-l2"><a class="reference internal" href="topical-low-level-ops.html">Low-level operations</a></li>
<li class="toctree-l2"><a class="reference internal" href="topical-bam.html">Working with BAM files</a></li>
<li class="toctree-l2"><a class="reference internal" href="topical-bam-semantics.html">Notes on BAM file semantics</a></li>
<li class="toctree-l2 current"><a class="current reference internal" href="#">Specifying genomes</a></li>
<li class="toctree-l2"><a class="reference internal" href="topical-random.html">Randomization</a></li>
<li class="toctree-l2"><a class="reference internal" href="topical-wrapping.html">Wrapping new tools</a></li>
<li class="toctree-l2"><a class="reference internal" href="topical-comparisons.html">Comparisons</a></li>
<li class="toctree-l2"><a class="reference internal" href="sh-comparison.html">Shell script comparison</a></li>
<li class="toctree-l2"><a class="reference internal" href="pybedtools-dev-history.html"><code class="xref py py-mod docutils literal notranslate"><span class="pre">pybedtools</span></code> development model</a></li>
<li class="toctree-l2"><a class="reference internal" href="flow-of-commands.html">Under the hood</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="FAQs.html">FAQs</a></li>
<li class="toctree-l1"><a class="reference internal" href="scripts.html">Scripts</a></li>
<li class="toctree-l1"><a class="reference internal" href="autodoc_source.html"><code class="xref py py-mod docutils literal notranslate"><span class="pre">pybedtools</span></code> Reference</a></li>
<li class="toctree-l1"><a class="reference internal" href="changes.html">Changelog</a></li>
</ul>

<div class="relations">
<h3>Related Topics</h3>
<ul>
  <li><a href="index.html">Documentation overview</a><ul>
  <li><a href="topical-documentation-contents.html">Topical Documentation</a><ul>
      <li>Previous: <a href="topical-bam-semantics.html" title="previous chapter">Notes on BAM file semantics</a></li>
      <li>Next: <a href="topical-random.html" title="next chapter">Randomization</a></li>
  </ul></li>
  </ul></li>
</ul>
</div>
<div id="searchbox" style="display: none" role="search">
  <h3 id="searchlabel">Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" aria-labelledby="searchlabel" autocomplete="off" autocorrect="off" autocapitalize="off" spellcheck="false"/>
      <input type="submit" value="Go" />
    </form>
    </div>
</div>
<script>document.getElementById('searchbox').style.display = "block"</script>








        </div>
      </div>
      <div class="clearer"></div>
    </div>

    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="topical-random.html" title="Randomization"
             accesskey="N">next</a> |</li>
        <li class="right" >
          <a href="topical-bam-semantics.html" title="Notes on BAM file semantics"
             accesskey="P">previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="index.html">pybedtools 0.9.0 documentation</a> &#187;</li>
          <li class="nav-item nav-item-1"><a href="topical-documentation-contents.html" accesskey="U">Topical Documentation</a> &#187;</li>
        <li class="nav-item nav-item-this"><a href="">Specifying genomes</a></li> 
      </ul>
    </div>


    <div class="footer">
      &copy;2010-2015, Ryan Dale.
      
      |
      Powered by <a href="http://sphinx-doc.org/">Sphinx 5.0.1</a>
      &amp; <a href="https://github.com/bitprophet/alabaster">Alabaster 0.7.12</a>
      
      |
      <a href="_sources/topical-genome.rst.txt"
          rel="nofollow">Page source</a>
    </div>

    

    
  </body>
</html>